create domain per request, so hub per request is created

jardakotesovec opened the pull request:
I suspect that creating `new Scope()` does not do anything.. its just empty object and it will stay empty. You can try - no breadcrumbs get recorded there.

After bit digging, I think only way to isolate requests is to have hub per request, which is saved in domain. This is pretty much what express middleware [does](https://github.com/getsentry/sentry-javascript/blob/master/packages/node/src/handlers.ts#L246). 

If you look at the [getCurrentHub](https://github.com/getsentry/sentry-javascript/blob/master/packages/hub/src/hub.ts#L393) it creates new hub in domain if there is active domain. Otherwise it uses one global hub. 

Using this would also add option to add some context anywhere within request lifecycle with 
```
Sentry.configureScope(scope => {
  scope.setTag('key', 'value');
});
```

All these global functions like `configureScope` or `withScope` they get hub [internally](https://github.com/getsentry/sentry-javascript/blob/master/packages/minimal/src/index.ts#L9) via `getCurrentHub` which gets the current hub from active domain, therefore should be scoped to the request.

I will do some testing sometime soon, but interested in your feedback. This code is mainly just to get point across for now.

Hi @kamilogorek if you would have few minutes just to glance over this PR if it makes sense? It would be super helpful :-) Just trying to make the hapi integration right.

fiws commented on index.js:
```diff
@@ -33,7 +34,8 @@ exports.register = (server, options) => {
     type: 'onRequest',
     method(request, h) {
-      request.sentryScope = new Sentry.Scope();
+      Domain.create();
```
Domain is not defined. I think this should be `domain`

see the test logs: https://travis-ci.com/hydra-newmedia/hapi-sentry/jobs/205971090

jardakotesovec replied:
Yep, it was too illustrative.  Now it should be also working..  will do some testing now.

jardakotesovec replied:
Did one more update, create() was not enough to active the domain. Now it creates the domain and its giving me breadcrumps only for that request, which is good.

fiws replied:
The test still fail though. You can just run `npm test` locally. you don't have to push. I don't mind if this code is not up to date and broken.

jardakotesovec replied:
@fiws yep, was trying to figure out how it should be done and wanted to share it early in case you have some insights/feedback. Tests are now passing.

jardakotesovec commented:
Ok, rebased. Added some comments and domain clean up. Also removed [WIP] as now I think its legitimate PR :-).  Removed test was testing sentryScope which is not used anymore, so removed that.

jardakotesovec commented:
In case you decide to go forward with this change, it probably should be major version bump, since that sentryScope is gone, which theoretically someone could use. Also anyone using domains should be aware of that update..  to basically ensure its not interfering with their use case in some bad way. Also unlikely, but possible.

guischdi commented:
What about https://github.com/getsentry/sentry-javascript/blob/master/packages/node/src/handlers.ts#L247-L255?
Don't we need something similar here to ensure the domain recognizes errors thrown in a hapi request handler? And another (maybe stupid) question: How is the Hub interconnected to the domain?

jardakotesovec commented:
@guischdi Hi, thanks for looking into it. 
My understanding why express integration has this approach is that domains are useful also for catching programmer and other errors - simply exceptions that don't end up following callback `next(err)` and would end up as unhandled exception without any context.  We don't care about that in hapi, since version 17 - domains where removed from there because its now relying on catches and promise chains - you can see more details in [changelog](https://github.com/hapijs/hapi/issues/3658).

So with this assumption - if you want to catch such error in domain and still have context which request caused it - thats exactly what they are doing on these lines. With hapi its fine to rely on that request error event and report errors there.

And with regards how the hub is interconnected with domain. Thats not stupid question, it took me sometime to understand that. Best is to look at the [`getCurrentHub`](https://github.com/getsentry/sentry-javascript/blob/master/packages/hub/src/hub.ts#L393), where you see how its looking for current hub. Point of domain is that it should be around for complete async chain of calls therefore for complete lifecycle of request. Hence for example if console breadcrump integration wants to add breadcrumb - it calls `getCurrentHub`, which checks if there is currently active domain. And either get existing hub from there or creates new one if its not there. So these hubs lives in individual domains. 

Hope it helps. And feel free to question it.. I might be also missing something.

guischdi commented:
Thanks @jardakotesovec for your extensive explanation. I think I now understand it.

Your approach seems to be working: I only get breadcrumbs of the specific request and global actions like `console.log`s, even if there are concurrent requests.

The thing about the global actions is, that you sometimes might have other global jobs like health checks running from time to time. This jobs might manipulate the scope, which is then also used to be attached to a sentry event. You can find sample code to reproduce this behavior here:
```javascript
const hapi = require('hapi');
const plugin = require('./');
const Sentry = require('@sentry/node');

const dsn = '<dsn of a test project>';

const delay = t => new Promise(resolve => setTimeout(resolve, t));

(async () => {
  const server = hapi.server({ port: 8080 });

  server.route({
    method: 'GET',
    path: '/test-sentry',
    async handler() {
      Sentry.configureScope(scope => scope.setTag('foo', 'bar'));
      console.log('Tag set to bar');
      await delay(2000);
      throw new Error('Oh no, bar is now 42!');
    },
  });

  await server.register({
    plugin,
    options: {
      client: { dsn },
    },
  });

  await server.start();
  console.log('Server started');

  server.inject({ method: 'GET', url: '/test-sentry' });

  await delay(1000);
  Sentry.configureScope(scope => scope.setTag('foo', 42));
})();
```

Sentry gets the tag value 42 which is not set in the request handler but in some globally running job, as can be seen here:
![sentry42](https://user-images.githubusercontent.com/14920913/59099068-6b15e080-8922-11e9-861a-7df72b86c6c4.png)


Maybe the deprecation warning (also to be found in the Sentry breadcrumbs of the event which was captured at first request made to the test server) gives you some hints how to fix that issue. Or maybe @kamilogorek has some advice how to limit the access to a Sentry hub (or is it only the scope which is accessible from somewhere else) to a single request.

jardakotesovec commented:
Can look at it more in detail on monday. But basically if there is no
active domain, which is the case in your ‘global’ example than it talks to
global hub and the hubs created in domain (within request) are somewhat
inherited from the global hub, so you can see it there.
If you want to report something to sentry which is specific for one event
you can always use that ‘withScope’ function and scope changes inside that
won’t leak to any other scope. Hope it somewhat answers your question.

pá 7. 6. 2019 v 12:56 odesílatel Christian Hotz <notifications@github.com>
napsal:

> Thanks @jardakotesovec <https://github.com/jardakotesovec> for your
> extensive explanation. I think I now understand it.
>
> Your approach seems to be working: I only get breadcrumbs of the specific
> request and global actions like console.logs, even if there are
> concurrent requests.
>
> The thing about the global actions is, that you sometimes might have other
> global jobs like health checks running from time to time. This jobs might
> manipulate the scope, which is then also used to be attached to a sentry
> event. You can find sample code to reproduce this behavior here:
>
> const hapi = require('hapi');const plugin = require('./');const Sentry = require('@sentry/node');
> const dsn = '<dsn of a test project>';
> const delay = t => new Promise(resolve => setTimeout(resolve, t));
>
> (async () => {
>   const server = hapi.server({ port: 8080 });
>
>   server.route({
>     method: 'GET',
>     path: '/test-sentry',
>     async handler() {
>       Sentry.configureScope(scope => scope.setTag('foo', 'bar'));
>       console.log('Tag set to bar');
>       await delay(2000);
>       throw new Error('Oh no, bar is now 42!');
>     },
>   });
>
>   await server.register({
>     plugin,
>     options: {
>       client: { dsn },
>     },
>   });
>
>   await server.start();
>   console.log('Server started');
>
>   server.inject({ method: 'GET', url: '/test-sentry' });
>
>   await delay(1000);
>   Sentry.configureScope(scope => scope.setTag('foo', 42));
> })();
>
> Sentry gets the tag value 42 which is not set in the request handler but
> in some globally running job, as can be seen here:
> [image: sentry42]
> <https://user-images.githubusercontent.com/14920913/59099068-6b15e080-8922-11e9-861a-7df72b86c6c4.png>
>
> Maybe the deprecation warning (also to be found in the Sentry breadcrumbs
> of the event which was captured at first request made to the test server)
> gives you some hints how to fix that issue. Or maybe @kamilogorek
> <https://github.com/kamilogorek> has some advice how to limit the access
> to a Sentry hub (or is it only the scope which is accessible from somewhere
> else) to a single request.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/hydra-newmedia/hapi-sentry/pull/21?email_source=notifications&email_token=AAMELGLNTCEIVNRZXTY6NC3PZI5FBA5CNFSM4HVBLQUKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODXFQBQA#issuecomment-499843264>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAMELGK2FOCA5Z3M2SLECELPZI5FBANCNFSM4HVBLQUA>
> .
>

jardakotesovec commented:
And I think any global actions are better of separated in own process run
by cron or something. And if you want them run in same process you can just
create and enter the domain at the beggining (or run it inside domain.run)
to get same separation as requests have.

pá 7. 6. 2019 v 13:41 odesílatel Jarda Kotěšovec <jarda.kotesovec@gmail.com>
napsal:

> Can look at it more in detail on monday. But basically if there is no
> active domain, which is the case in your ‘global’ example than it talks to
> global hub and the hubs created in domain (within request) are somewhat
> inherited from the global hub, so you can see it there.
> If you want to report something to sentry which is specific for one event
> you can always use that ‘withScope’ function and scope changes inside that
> won’t leak to any other scope. Hope it somewhat answers your question.
>
> pá 7. 6. 2019 v 12:56 odesílatel Christian Hotz <notifications@github.com>
> napsal:
>
>> Thanks @jardakotesovec <https://github.com/jardakotesovec> for your
>> extensive explanation. I think I now understand it.
>>
>> Your approach seems to be working: I only get breadcrumbs of the specific
>> request and global actions like console.logs, even if there are
>> concurrent requests.
>>
>> The thing about the global actions is, that you sometimes might have
>> other global jobs like health checks running from time to time. This jobs
>> might manipulate the scope, which is then also used to be attached to a
>> sentry event. You can find sample code to reproduce this behavior here:
>>
>> const hapi = require('hapi');const plugin = require('./');const Sentry = require('@sentry/node');
>> const dsn = '<dsn of a test project>';
>> const delay = t => new Promise(resolve => setTimeout(resolve, t));
>>
>> (async () => {
>>   const server = hapi.server({ port: 8080 });
>>
>>   server.route({
>>     method: 'GET',
>>     path: '/test-sentry',
>>     async handler() {
>>       Sentry.configureScope(scope => scope.setTag('foo', 'bar'));
>>       console.log('Tag set to bar');
>>       await delay(2000);
>>       throw new Error('Oh no, bar is now 42!');
>>     },
>>   });
>>
>>   await server.register({
>>     plugin,
>>     options: {
>>       client: { dsn },
>>     },
>>   });
>>
>>   await server.start();
>>   console.log('Server started');
>>
>>   server.inject({ method: 'GET', url: '/test-sentry' });
>>
>>   await delay(1000);
>>   Sentry.configureScope(scope => scope.setTag('foo', 42));
>> })();
>>
>> Sentry gets the tag value 42 which is not set in the request handler but
>> in some globally running job, as can be seen here:
>> [image: sentry42]
>> <https://user-images.githubusercontent.com/14920913/59099068-6b15e080-8922-11e9-861a-7df72b86c6c4.png>
>>
>> Maybe the deprecation warning (also to be found in the Sentry breadcrumbs
>> of the event which was captured at first request made to the test server)
>> gives you some hints how to fix that issue. Or maybe @kamilogorek
>> <https://github.com/kamilogorek> has some advice how to limit the access
>> to a Sentry hub (or is it only the scope which is accessible from somewhere
>> else) to a single request.
>>
>> —
>> You are receiving this because you were mentioned.
>>
>>
>> Reply to this email directly, view it on GitHub
>> <https://github.com/hydra-newmedia/hapi-sentry/pull/21?email_source=notifications&email_token=AAMELGLNTCEIVNRZXTY6NC3PZI5FBA5CNFSM4HVBLQUKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODXFQBQA#issuecomment-499843264>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AAMELGK2FOCA5Z3M2SLECELPZI5FBANCNFSM4HVBLQUA>
>> .
>>
>

guischdi commented:
I didn't say, I WANT to access a scope used (in a request) from somewhere else. I meant, one COULD do so. So my concern was not about whether changes on a scope in a request handler could leak to some other part of the application. It's about the other direction: No matter if I use `withScope` or `configureScope` the global operation (like a database health check or something) could accidentally overwrite some scope properties, which will be passed to sentry by some completely other part of the application (the request handler).

jardakotesovec commented:
@guischdi Yes, I understand that. My thoughts on this particular scenario: 
- it should not happen with `withScope`, only with `configureScope`
- it should not overwrite things in hapi requests, "only" leak/extend. If you are setting same field globally and in request, request always win. Did not test that, but after hub is inherited for hapi request, than you can't touch it from the global hub.
- you can isolate your global health checks with domain, which would prevent that, but you have to be aware how the sentry works
- you could trigger your async tasks with server.inject and do them inside hapi handlers, so you get all the hapi error handling and domains without much effort

But you are right, something from global hub still can get inside requests, which in some cases might be intentional (like some global information about that node instance), but also can be unexpected in other use cases.  

On the other hand before this change it was significantly more mixed up and noone really complaint, so its not probably that big deal :-). And I think its also good to consider that domains are already pretty magic (one day will be replaced with async hooks in sentry) - when running things in one thread, its not easy to not mix up things without just passing them everywhere as arguments.

kamilogorek commented:
Because of the way you are plugging in into the hapi error handlers, you can do it in 2 ways, which doesn't require domains.

a) Hub based: create it and attach it to the request itself:

```js
server.ext({
    type: 'onRequest',
    method(request, h) {
      const currentHub = Sentry.getCurrentHub()
      const requestHub = new Sentry.Hub(
        currentHub.getClient(),
        Sentry.Scope.clone(currentHub.getScope())
      )
      Sentry.setHubOnCarrier(request, requestHub);
      return h.continue;
    },
});

// and then

server.events.on({ name: 'request', channels: opts.channels }, async (request, event) => {
  const hub = Sentry.getHubFromCarrier(request)

  hub.withScope(scope => {
    scope.addEventProcessor(_sentryEvent => {
      // do whatever
    });

    Hoek.applyToDefaults(scope, request.sentryScope);
    hub.captureException(event.error);
  });
});
```

b) Client/Scope based: use client directly and give it manually handled scope

```js
server.ext({
    type: 'onRequest',
    method(request, h) {
      // Inherit data from the parent
      request.sentryScope = Sentry.Scope.clone(currentHub.getScope());
      return h.continue;
    },
});

// and then

server.events.on({ name: 'request', channels: opts.channels }, async (request, event) => {
  const client = Sentry.getCurrentHub().getClient()
  const scope = request.sentryScope;

  scope.addEventProcessor(_sentryEvent => {
    // do whatever
  });
  // or set directly
  scope.setExtra("what", "ever")

  Hoek.applyToDefaults(scope, request.sentryScope);
  client.captureException(event.error, {
      // this is EventHint, so pass it whatever you want to expose
      request,
      event
  }, scope);
});
```

jardakotesovec commented:
@kamilogorek Thanks a lot for getting back to us. Attaching hub to req was option I considered, but was inclined to use domains because thats only way to also save breadcrumps and other integrations (which are using getCurrentHub) to request hub.

Only reason why not to use domains I would think are potential performance/stability issues, but since its used a lot inside sentry I would think you don't find domains troublesome?

kamilogorek commented:
@jardakotesovec you're right, and it should be totally fine to use them.
However, I'm not 100% sure how Hapi works, so I'd write some manual tests around it as well, see express tests for the reference - https://github.com/getsentry/sentry-javascript/blob/master/packages/node/test/manual/express-scope-separation/start.js

guischdi commented:
Thanks @kamilogorek for your responses and having a look into the PR implementation!

@jardakotesovec I understand, why using domains seems the right way of doing it.
But maybe you could do two things for getting this PR to our utmost satisfaction:
1. Write some test analogous to the linked express scope separation test.
1. Adjust the "Scope" section in README.md to the new way of handling scopes from route handlers.

jardakotesovec commented:
@guischdi Yes, tests are always nice to have. I am on holidays for rest of the week and next week, hopefully I will be able to find some spare time to make progress on that then.

mkg20001 commented:
@jardakotesovec You might want to rebase this against latest master _after_ writing tests, so you can check if the new deps break anything

guischdi commented:
Sorry for the latency, @jardakotesovec !
I rebased your branch, wrote some tests and re-enabled our scope handling (which should have nothing to do with your introduced changes).

You may have a look at this. If everything is fine with you, we would release this changes with a new semver major (due to dependency updates performed in the meantime). I introduced a new option (`useDomainPerRequest`) with default value `false`, because domains imply [performance issues](https://nodejs.org/en/docs/guides/domain-postmortem/#performance-issues).

fiws commented on package.json:
```diff
@@ -32,7 +32,8 @@
     "@hapi/joi": "^15.1.0",
-    "@sentry/node": "^5.4.3"
+    "@sentry/node": "^5.4.3",
+    "delay": "^4.3.0"
```
@guischdi this is a dev dependency i think

jardakotesovec commented:
@guischdi Thanks! Looks good. Sorry for not responding I was on holidays.
